### PR TITLE
[nhcb branch] Use single bit to differentiate between optimized bounds and floats

### PR DESCRIPTION
**Pull request by krajorama**
Use one bit to decide what kind of data to read/write. This reduces storage need of floats from 72 bits to 65 bits and makes the integers store in 17 instead of 16.

### CODE REVIEW

**Comment by bboreham** [comment]
What is the `nhcb` branch?  Is it ok that `TestQueryStatistics` failed in CI?

Doesn't this make a breaking change to the storage on disk?

**Comment by krajorama** [comment]
> What is the `nhcb` branch? Is it ok that `TestQueryStatistics` failed in CI?
> 
> Doesn't this make a breaking change to the storage on disk?

What it is:
https://github.com/prometheus/proposals/blob/main/proposals/2024-01-26_classic-histograms-stored-as-native-histograms.md
EPIC: https://github.com/prometheus/prometheus/issues/13485

It's not a breaking change, doesn't affect histogram chunk with the current schema for exponential histograms.

**Comment by beorn7** [comment]
Generally, I would say it helps a bit if we store things byte-aligned. Of course, it depends on the exact location in the bit stream if things are byte aligned at the position where we store the NHCB bounds. If generally not byte-aligned stuff has been in the stream before anyway, maybe we could also use the varbit (not varint) encoding for integers and be even a bit more compact.
(I preferred varint previously because I somehow assumed we are still byte-aligned, but that might actually not be true. Needs verification.)

**Comment by beorn7** [comment]
OK, I checked. 

We are definitely not byte-aligned at the time `putHistogramChunkLayoutCustomBounds` is called.

So I would like to modify the algorithm once again:

- If you get a positive integer number (or zero) after multiplying with 1000, write a 1 bit and then use `putVarbitUint` to write that integer.
- Otherwise, do what's done in this PR (write a 0 bit and the 64 bit of the float64).

**Comment by zenador** [comment]
Actually, looking at how varbit is implemented there, maybe this optimisation could work? https://github.com/zenador/prometheus/commit/1f97f359f88a0674940563970984abc1f8a4b6fe

And once the implementation is finalised, we'd need to update the doc comments.

**Comment by beorn7** [comment]
> Actually, looking at how varbit is implemented there, maybe this optimisation could work? [zenador@1f97f35](https://github.com/zenador/prometheus/commit/1f97f359f88a0674940563970984abc1f8a4b6fe)

Yes, this should work.

**Comment by krajorama** [comment]
> > Actually, looking at how varbit is implemented there, maybe this optimisation could work? [zenador@1f97f35](https://github.com/zenador/prometheus/commit/1f97f359f88a0674940563970984abc1f8a4b6fe)
> 
> Yes, this should work.

Picked @zenador optimization, nice find!

File: tsdb/chunkenc/histogram_meta.go [hunk 1]
```diff
@@ -195,17 +195,18 @@ func isWholeWhenMultiplied(in float64) bool {
...
-	if tf < 0 || tf > 16382 || !isWholeWhenMultiplied(f) {
-		b.putUvarint(0)
+	// 33554431-1 comes from the maximum that can be stored in a varint in 4 bytes, other values are stored in 8 bytes anyway.
+	if tf < 0 || tf > 33554430 || !isWholeWhenMultiplied(f) {
+		b.writeBit(zero)
 		b.writeBits(math.Float64bits(f), 64)
 		return
 	}
-	b.putUvarint(uint64(math.Round(tf) + 1))
+	putVarbitUint(b, uint64(math.Round(tf)+1))
```
[the remark this answers is not in the document]
Comment (krajorama) [reply tsdb/chunkenc/histogram_meta.go:204]: And more symmetrical

**Comment by zenador** [comment]
LGTM

**Pull request merged**